In [8]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import functions as F
from pyspark import SQLContext
import spark
from pyspark.sql import types

In [2]:
conf = pyspark.SparkConf()
spark = SparkSession.builder.config(conf=conf).getOrCreate()
#sqlc = SQLContext(sc)

In [3]:
sc = spark.sparkContext
sc.setLogLevel("INFO")

In [4]:
sqlc = SQLContext(sc)

/usr/local/Cellar/apache-spark/3.3.1/libexec/python/pyspark/sql/context.py:112: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(


# Q1 - Spark Version

In [5]:
spark.version

'3.3.1'

# Q2 - HVFHW June 2021 

## download the data

In [6]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz

--2023-02-26 17:55:18--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Auflösen des Hostnamens github.com (github.com)… 140.82.121.3
Verbindungsaufbau zu github.com (github.com)|140.82.121.3|:443 … verbunden.
HTTP-Anforderung gesendet, auf Antwort wird gewartet … 302 Found
Platz: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230226%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230226T165518Z&X-Amz-Expires=300&X-Amz-Signature=5b118b941f4776b19c7263ebafd800361f9df2f1d5f67e830b6477c5c2c958e5&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [folgend]
--2023-02-26 17:55:18--  https://objects.githubusercontent.com/github-production-release-ass

In [7]:
!ls

2023_02_26_Pyspark_week5_homework.ipynb fhvhv_tripdata_2021-06.csv.gz


In [21]:
!gunzip fhvhv_tripdata_2021-06.csv.gz

In [22]:
!ls

2023_02_26_Pyspark_week5_homework.ipynb fhvhv_tripdata_2021-06.csv


In [23]:
!head fhvhv_tripdata_2021-06.csv

dispatching_base_num,pickup_datetime,dropoff_datetime,PULocationID,DOLocationID,SR_Flag,Affiliated_base_number
B02764,2021-06-01 00:02:41,2021-06-01 00:07:46,174,18,N,B02764
B02764,2021-06-01 00:16:16,2021-06-01 00:21:14,32,254,N,B02764
B02764,2021-06-01 00:27:01,2021-06-01 00:42:11,240,127,N,B02764
B02764,2021-06-01 00:46:08,2021-06-01 00:53:45,127,235,N,B02764
B02510,2021-06-01 00:45:42,2021-06-01 01:03:33,144,146,N,
B02510,2021-06-01 00:18:15,2021-06-01 00:25:47,49,17,N,
B02510,2021-06-01 00:33:06,2021-06-01 00:42:46,49,225,N,
B02510,2021-06-01 00:46:27,2021-06-01 00:56:50,225,177,N,
B02764,2021-06-01 00:48:06,2021-06-01 01:04:10,209,45,N,B02764


In [24]:
schema_fhvhv = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True)
])

In [26]:
fhvhv_2021_6 = sqlc.read \
    .option("header", "true") \
    .schema(schema_fhvhv) \
    .csv('fhvhv_tripdata_2021-06.csv')

In [27]:
fhvhv_2021_6.count()

14961892

In [28]:
fhvhv_2021_6.dtypes

[('dispatching_base_num', 'string'),
 ('pickup_datetime', 'timestamp'),
 ('dropoff_datetime', 'timestamp'),
 ('PULocationID', 'int'),
 ('DOLocationID', 'int'),
 ('SR_Flag', 'string'),
 ('Affiliated_base_number', 'string')]

In [29]:
fhvhv_2021_6.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'SR_Flag',
 'Affiliated_base_number']

In [30]:
fhvhv_2021_6.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02764|2021-06-01 00:02:41|2021-06-01 00:07:46|         174|          18|      N|                B02764|
|              B02764|2021-06-01 00:16:16|2021-06-01 00:21:14|          32|         254|      N|                B02764|
|              B02764|2021-06-01 00:27:01|2021-06-01 00:42:11|         240|         127|      N|                B02764|
|              B02764|2021-06-01 00:46:08|2021-06-01 00:53:45|         127|         235|      N|                B02764|
|              B02510|2021-06-01 00:45:42|2021-06-01 01:03:33|         144|         146|      N|                  null|
|              B02510|2021-06-01 00:18:1

In [31]:
fhvhv_2021_6_repartition = fhvhv_2021_6.repartition(12)

In [32]:
fhvhv_2021_6_repartition.write.parquet('fhvhv_2021_6',mode='overwrite')

In [ ]:
!ls -lh fhvhv_2021_6

23MB

# Q3 How many taxi trips were there on June 15?

In [37]:
fhvhv_2021_6_repartition = fhvhv_2021_6_repartition.withColumn("pickup_date",F.to_date("pickup_datetime"))

In [38]:
fhvhv_2021_6_repartition.where("pickup_date = '2021-06-15'").count()

452470

# Q4 Now calculate the duration for each trip. How long was the longest trip in Hours?

In [47]:
fhvhv_2021_6_repartition = fhvhv_2021_6_repartition.withColumn("duration",
                        F.round((F.col("dropoff_datetime")-F.col("pickup_datetime")).cast("long")/60.0/60.0,2))

In [48]:
fhvhv_2021_6_repartition.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+--------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|pickup_date|duration|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+--------+
|              B02869|2021-06-04 17:46:35|2021-06-04 18:05:56|         242|         213|      N|                B02869| 2021-06-04|    0.32|
|              B02871|2021-06-01 20:16:14|2021-06-01 20:34:29|         214|         206|      N|                B02871| 2021-06-01|     0.3|
|              B02617|2021-06-04 19:34:16|2021-06-04 20:09:29|         142|         144|      N|                B02617| 2021-06-04|    0.59|
|              B02880|2021-06-04 16:20:12|2021-06-04 18:11:36|         189|         265|      N|                B02880| 2021-06-04|    1.86|
|            

In [49]:
fhvhv_2021_6_repartition.orderBy("duration",ascending=False).select("pickup_datetime","dropoff_datetime","duration").show()

+-------------------+-------------------+--------+
|    pickup_datetime|   dropoff_datetime|duration|
+-------------------+-------------------+--------+
|2021-06-25 13:55:41|2021-06-28 08:48:25|   66.88|
|2021-06-22 12:09:45|2021-06-23 13:42:44|   25.55|
|2021-06-27 10:32:29|2021-06-28 06:31:20|   19.98|
|2021-06-26 22:37:11|2021-06-27 16:49:01|    18.2|
|2021-06-23 20:40:43|2021-06-24 13:08:44|   16.47|
|2021-06-23 22:03:31|2021-06-24 12:19:39|   14.27|
|2021-06-24 23:11:00|2021-06-25 13:05:35|   13.91|
|2021-06-04 20:56:02|2021-06-05 08:36:14|   11.67|
|2021-06-27 07:45:19|2021-06-27 19:07:16|   11.37|
|2021-06-20 17:05:12|2021-06-21 04:04:16|   10.98|
|2021-06-01 12:25:29|2021-06-01 22:41:32|   10.27|
|2021-06-01 12:01:46|2021-06-01 21:59:45|    9.97|
|2021-06-28 13:13:59|2021-06-28 23:11:58|    9.97|
|2021-06-27 03:52:14|2021-06-27 13:30:30|    9.64|
|2021-06-18 08:50:29|2021-06-18 18:27:57|    9.62|
|2021-06-08 16:38:14|2021-06-09 02:07:03|    9.48|
|2021-06-11 23:26:20|2021-06-12

on 2021-06-25 13:55:41 with 66.88 hours

# Q5 User Interface

opening localhost:4040

# Q6 Most frequent pickup location zone

In [50]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2023-02-26 18:22:58--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Auflösen des Hostnamens github.com (github.com)… 140.82.121.3
Verbindungsaufbau zu github.com (github.com)|140.82.121.3|:443 … verbunden.
HTTP-Anforderung gesendet, auf Antwort wird gewartet … 302 Found
Platz: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230226%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230226T172258Z&X-Amz-Expires=300&X-Amz-Signature=fde9476406cec1d2ee1f5c6e30a91015bd39d299724f10e2ee91d89e9193098b&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [folgend]
--2023-02-26 18:22:58--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948

In [58]:
schema_zones = types.StructType([
    types.StructField('LocationID', types.IntegerType(), True),
    types.StructField('Borough', types.StringType(), True),
    types.StructField('Zone', types.StringType(), True),
    types.StructField('service_zone', types.StringType(), True),
])

In [59]:
zones = sqlc.read \
    .option("header", "true") \
    .schema(schema_zones) \
    .csv('taxi_zone_lookup.csv')

In [60]:
zones.columns

['LocationID', 'Borough', 'Zone', 'service_zone']

In [61]:
zones.schema

StructType([StructField('LocationID', IntegerType(), True), StructField('Borough', StringType(), True), StructField('Zone', StringType(), True), StructField('service_zone', StringType(), True)])

In [62]:
zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [67]:
fhvhv_2021_6_repartition_zones = fhvhv_2021_6_repartition.join(
        zones.select("LocationID","Zone"),
        fhvhv_2021_6_repartition.PULocationID == zones.LocationID,
        "left").withColumnRenamed("Zone","pickup_zone").drop("LocationID")\
        .join(
        zones.select("LocationID","Zone"),
        fhvhv_2021_6_repartition.DOLocationID == zones.LocationID,
        "left").withColumnRenamed("Zone","dropoff_zone").drop("LocationID")

In [69]:
fhvhv_2021_6_repartition_zones.groupBy("pickup_zone").count().orderBy("count",ascending=False).show()

+--------------------+------+
|         pickup_zone| count|
+--------------------+------+
| Crown Heights North|231279|
|        East Village|221244|
|         JFK Airport|188867|
|      Bushwick South|187929|
|       East New York|186780|
|TriBeCa/Civic Center|164344|
|   LaGuardia Airport|161596|
|            Union Sq|158937|
|        West Village|154698|
|             Astoria|152493|
|     Lower East Side|151020|
|        East Chelsea|147673|
|Central Harlem North|146402|
|Williamsburg (Nor...|143683|
|          Park Slope|143594|
|  Stuyvesant Heights|141427|
|        Clinton East|139611|
|West Chelsea/Huds...|139431|
|             Bedford|138428|
|         Murray Hill|137879|
+--------------------+------+
only showing top 20 rows



Crown Heights North with 231'279 pickups